doc/cl_status.1
.TH CL_STATUS 1 "12th May 2004"
.SH NAME
cl_status \- Check Status of the High-Availability Linux (Linux-HA) Subsystem
.SH SYNOPSIS
.B cl_status
.nh
.RI "<sub-command>" "[<options>] [<parameters>]"
.P
.SH DESCRIPTION
\fBcl_status\fP is used to check the status of the High-Availability Linux Subsystem

.SH SUB-COMMANDS
.TP
\fBhbstatus\fP
Indicate if heartbeat is running on the local system.
.TP
\fBlistnodes\fP
List the nodes in the cluster
.TP
\fBnodetype\fP \fITYPE\fP
List the nodes of the given type.
Currently the available types are ping and normal.
.TP
\fBlisthblinks\fP \fINODENAME\fP
List the network interfaces used as heartbeat links.
Nodename should be as per the ha.cf file for the cluster.
.TP
\fBhblinkstatus\fP \fINODENAME LINKNAME\fP
Show the status of a heartbeat link.
Nodename should be as per the ha.cf file for the cluster.
Linkname should be as per the output of the listhblinks subcomand.
.TP
\fBclientstatus\fP \fINODENAME CLIENTID [TIMEOUT]\fP
Show the status of heartbeat clients. 
Nodename and Client-ID should be as per the ha.cf file for the cluster.
Timeout is in milliseconds, the default is 100ms.
.TP
\fBrscstatus\fP
Show the status of cluster resources.
Status will be one of: local, foreign, all or none.
.TP
\fBhbparameter -p\fP \fIPARAMETERNAME\fP
Retrieve the value of cluster parameters.
The parameters may be one of the following, however some parameters
which have not been specified may not retrieved:
.RS 1i
apiauth,
auto_failback,
baud,
debug,
debugfile,
deadping,
deadtime,
hbversion,
hopfudge,
initdead,
keepalive,
logfacility,
logfile,
msgfmt,
nice_failback,
node,
normalpoll,
stonith,
udpport,
warntime,
watchdog
.RE

.SH OPTIONS
.TP
\fB-m\fP
Make the output more human readable.
The default output should be easier for scripts to parse.
Available with all commands.
.TP
\fB-p\fP
List only 'ping' nodes.
Available with hbstatus sub-command.
.TP
\fB-n\fP
List only 'normal' nodes.
Available with hbstatus sub-command.

.SH SEE ALSO
heartbeat(8)

.SH DOCUMENTATION
More information may be found at
.UR http://www.linux-ha.org/
http://www.linux-ha.org/
.UE

.SH AUTHORS

cl_status - Sun Jiang Dong
.UR mailto:sunjd@cn.ibm.com
<sunjd@cn.ibm.com>
.br
.UE
linux-ha - Alan Robertson
.UR mailto:alanr@unix.sh
<alanr@unix.sh>
.UE
\fIet al\fP.
.br
man page - Horms
.UR mailto:horms@verge.net.au
<horms@verge.net.au>
.UE



ChangeLog

* Wed Jun 23 2010 Andrew Beekhof <andrew@beekhof.net> 1.0.9.1-1
- Update source tarball to revision: c9a5915827d0 (stable-1.0) tip
- Statistics:
  Changesets: 4
  Diff:       2 files changed, 3 insertions(+), 1 deletion(-)

- Changes since Pacemaker-1.0.9
  + High: ais: Fix unterminated recursion in init_ais_connection()

* Wed Jun 23 2010 Andrew Beekhof <andrew@beekhof.net> 1.0.9-1
- Update source tarball to revision: 0796085f3833+ (stable-1.0) tip
- Statistics:
  Changesets: 152
  Diff:       266 files changed, 14324 insertions(+), 3842 deletions(-)

- Changes since Pacemaker-1.0.8
  + High: ais: Ensure the list of active processes sent to clients is always up-to-date
  + High: ais: Fix previous commit, actually return a result in get_process_list()
  + High: ais: Fix two more uses of getpwnam() in non-thread-safe locations
  + High: ais: Look for the correct conf variable for turning on file logging
  + High: ais: Need to find a better and thread-safe way to set core_uses_pid. Disable for now.
  + High: ais: Use the threadsafe version of getpwnam
  + High: cib: Also free query result for xpath operations that return more than one hit
  + High: cib: Fix the application of unversioned diffs
  + High: cib: Remove old developmental error logging
  + High: Core: Bug lf#2414 - Prevent use-after-free reported by valgrind when doing xpath based deletions
  + High: Core: Fix memory leak in replace_xml_child() reported by valgrind
  + High: Core: fix memory leaks exposed by valgrind
  + High: crmd: Bug 2401 - Improved detection of partially active peers
  + High: crmd: Bug lf#2379 - Ensure the cluster terminates when the PE is not available
  + High: crmd: Bug lf#2414 - Prevent use-after-free of the PE connection after it dies
  + High: crmd: Bug lf#2439 - cancel_op() can also return HA_RSCBUSY
  + High: crmd: Bug lf#2439 - Handle asynchronous notification of resource deletion events
  + High: crmd: Do not allow the target_rc to be misused by resource agents
  + High: crmd: Do not ignore action timeouts based on FSA state
  + High: crmd: Ensure we dont get stuck in S_PENDING if we loose an election to someone that never talks to us again
  + High: crmd: Fix memory leaks exposed by valgrind
  + High: crmd: Remove race condition that could lead to multiple instances of a clone being active on a machine
  + High: crmd: Send erase_status_tag() calls to the local CIB when the DC is fenced, since there is no DC to accept them
  + High: PE: Bug lf#1959 - Fail unmanaged resources should not prevent other services from shutting down
  + High: PE: Bug lf#2383 - Combine failcounts for all instances of an anonymous clone on a host
  + High: PE: Bug lf#2384 - Fix intra-set colocation and ordering
  + High: PE: Bug lf#2403 - Enforce mandatory promotion (colocation) constraints
  + High: PE: Bug lf#2412 - Correctly locate clone instances by their prefix
  + High: PE: Bug lf#2422 - Ordering dependencies on partially active groups not observed properly
  + High: PE: Bug lf#2424 - Use notify oepration definition if it exists in the configuration
  + High: PE: Bug lf#2433 - No services should be stopped until probes finish
  + High: PE: Do not be so quick to pull the trigger on nodes that are coming up
  + High: PE: Fix colocation for interleaved clones
  + High: PE: Fix colocation with partially active groups
  + High: PE: Fix memory leaks reported by valgrind
  + High: PE: Make the current data set a global variable so it does not need to be passed around everywhere
  + High: PE: Prevent endless loop when looking for operation definitions in the configuration
  + High: PE: Rewrite native_merge_weights() to avoid Fix use-after-free
  + High: Shell: always reload status if working with the cluster (bnc#590035)
  + High: Tools: crm_mon - fix memory leaks exposed by valgrind
  + Medium: ais: Correctly set logfile permissions in all cases
  + Medium: ais: create the final directory too for resource agents (bnc#603190)
  + Medium: ais: Make sure debug messages make it into the logfiles too
  + Medium: Build: Do not enable the -ansi compiler option by default, prevents use of strtoll()
  + Medium: cib: Bug lf#2352 - Changes to group order are not detected or broadcast to peers
  + Medium: cib: Correctly free the cib contents at signoff when in file-based mode
  + Medium: cib: xpath - Allow all hits to be deleted, allow the no_children option to return multiple hits
  + Medium: PE: Bug lf#2391 - Ensure important options (notify, unique, etc) are always exposed during resource operations
  + Medium: PE: Bug lf#2410 - Do not complain about missing agents during probes of a-symetric clusters
  + Medium: PE: Bug lf#2426 - stop-all-resources should not apply to stonith resources
  + Medium: PE: Bug lf#2435 - Support colocation sets with negative scores
  + Medium: PE: Check for use-of-NULL in dump_node_scores()
  + Medium: PE: Do not overwrite existing meta attributes (like timeout) for notify operations
  + Medium: PE: Ensure deallocated resources are stopped
  + Medium: PE: If there are no compatible peers when interleaving clones, ensure the instance is stopped
  + Medium: PE: Ignore colocation weights from clone instances
  + Medium: RA: SystemHealth: exit properly when the required software is not installed (bnc#587940)

tools/cl_status.c
/* File: cl_status.c
 * Description: 
 * 	A small tool for acquire the state information of heartbeat cluster.
 * TODO: Map string output to return value?
 *
 * Author: Sun Jiang Dong <sunjd@cn.ibm.com>
 * Copyright (c) 2004 International Business Machines
 * Referred to the following tools
 *	api_test Copyright (C) 2000 Alan Robertson <alanr@unix.sh>
 *	hbinfo   Copyright (C) 2004 Mike Neuhauser <mike@firmix.at>
 * 
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc., 59 Temple
 * Place - Suite 330, Boston, MA  02111-1307, USA.
 */

#include <portability.h>
#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>
#ifdef HAVE_GET_OPT_H
#include <getopt.h>
#endif
#include <stdarg.h>
#include <string.h>
#include <syslog.h>
#include <glib.h>
#include <clplumbing/cl_log.h>
#include <hb_api.h>

/* exit code */
const int OK = 0,
	  NORMAL_FAIL = 1,		/* such as the local node is down */
	  PARAMETER_ERROR = 11,
	  TIMEOUT = 12,
	  UNKNOWN_ERROR = 13;		/* error due to unkown causes */ 
/*
 * The exit values under some situations proposed by Alan.
 * nodestatus    fail when the node is down
 * clientstatus  fail when client not accessible (offline?)
 * hbstatus      fail when heartbeat not running locally
 * hblinkstatus  fail if the given heartbeat link is down
 * hbparameter   fail if the given parameter is not set to any value
*/

/*
 * Important
 * General return value for the following functions, and it is actually 
 * as this program cl_status' return value:
 * 	0(OK):   on success, including the node status is ok.
 * 	<>0: on fail
		1(NORMAL_FAIL):     for a "normal" failure (like node is down)
 * 		2(PARAMETER_ERROR):  
 *		3(OTHER_ERROR):	    error due to unkown causes
 */

/* 
 * Description: 
 * 	Detect if heartbeat is running.
 *
 * Parameters:
 *	Obvious. ;-) 
 *
 * Return Value:
 *	OK: 		In local machine, heartbeat is running.
 *	NORMAL_FAIL:    In local machine, heartbeat is stopped.
 */
static int
hbstatus(ll_cluster_t *hb, int argc, char ** argv, const char * optstr);

/*
 * Return Value:
 * 	OK:  	      In local machine, this operation succeed.
 *	NORMAL_FAIL:  In local machine, heartbeat is stopped.
 */
static int
listnodes(ll_cluster_t *hb, int argc, char ** argv, const char * optstr);

/*
 * Return Value:
 * 	OK: 	     the node is active.
 *	NORMAL_FAIL: the node is down
 */
static int
nodestatus(ll_cluster_t *hb, int argc, char ** argv, const char * optstr);

/*
 * Return Value:
 *	0: normal
 *	1: ping
 *	3: unknown type
 * Notes: not map string std_output to return value yet 
 */
static int
nodetype(ll_cluster_t *hb, int argc, char ** argv, const char * optstr);

/*
 * Return Value:
 *	0(OK):		sucess
 * 	1(NORMAL_FAIL): the node is down.
 */
static int
listhblinks(ll_cluster_t *hb, int argc, char ** argv, const char * optstr);

/*
 * Return Value:
 *	0(OK):		the link is up
 *	1(NORMAL_FAIL): the link is down
 */
static int
hblinkstatus(ll_cluster_t *hb, int argc, char ** argv, const char * optstr);

/*
 * Return Value:
 * 	0(OK): 		online
 *	1(NORMAL_FAIL): offline
 *	2:		join
 *	3: 		leave
 *
 *   When sucess and without -m option, at the meantime on stdout print one 
 *   of the following string to reflect the status of the client: 
 *	online, offline, join, leave
 */
static int
clientstatus(ll_cluster_t *hb, int argc, char ** argv, const char * optstr);

/*
 * Return Value:
 * 	0(OK): 		on success.
 *	1(NORMAL_FAIL): the node is down.
 *
 *   When sucess and without -m option, on stdout print one of the following
 *   string to reflect the status of the resource: 
 * 	none, local, foreign, all
 */
static int
rscstatus(ll_cluster_t *hb, int argc, char ** argv, const char * optstr);

/*
 * Return Value:
 *	0: success
 *	1: fail if the given parameter is not set to any value
 */
static int
hbparameter(ll_cluster_t *hb, int argc, char ** argv, const char * optstr);

/* miscellaneous functions */
static int test(ll_cluster_t *hb, int argc, char ** argv, const char * optstr);
static int general_simple_opt_deal(int argc, char ** argv, const char * optstr);

typedef struct {
	const char * name;
	int (*func)(ll_cluster_t *hb, int, char **, const char *);
	const char * optstr;
} cmd_t;

static const size_t CMDS_MAX_LENGTH = 16;
static gboolean FOR_HUMAN_READ = FALSE;
static const cmd_t cmds[] = {
	{ "hbstatus",      hbstatus, 	  "m" },
	{ "listnodes",     listnodes, 	  "mpn" },
	{ "nodestatus",    nodestatus, 	  "m" },
	{ "nodetype",      nodetype, 	  "m" },
	{ "listhblinks",   listhblinks,   "m" },
	{ "hblinkstatus",  hblinkstatus,  "m" },
	{ "clientstatus",  clientstatus,  "m" },
	{ "rscstatus",     rscstatus, 	  "m"}, 
	{ "hbparameter",   hbparameter,	  "mp:"},
	{ "test",	   test,	  NULL},
	{ NULL, NULL, NULL },
};

static const char * simple_help_screen =
"Usage: cl_status <sub-command> [<options>] [<parameters>]\n"
"\n"
"The sub-commands usage is showed as below.\n"
"Note: There is a general option -m for all subcommands, to make the output\n"
"      more human readable. The default output should be easier for scripts\n"
"      to parse.\n"
"\n"
"hbstatus\n"
"	Indicate if heartbeat is running on the local system.\n"
"\n"
"listnodes\n"
"	List the nodes in the cluster.\n"
"	Options:\n"
"	-p	list only 'ping' type nodes\n"
"	-n	list only 'normal' type nodes\n"
"\n"
"nodestatus <node-name>\n"
"	List the node status.\n"
"\n"
"nodetype <node-name>\n"
"	List the nodes of a given type.\n"
"       Curretly the available types are ping and normal.\n"
"\n"
"listhblinks <node-name>\n"
"	List the network interfaces used as hearbeat links.\n"
"\n"
"hblinkstatus <node-name> <link-name>\n"
"	Show the status of a heartbeat link\n"
"\n"
"clientstatus <node-name> <client-id> [<timeout>]\n"
"	Show the status of heartbeat clients.\n"
"       For example, ping or ipfail.\n"
"\n"
"rscstatus\n"
"	Show the status of cluster resources.\n"
"       Status will be one of: local, foreign, all or none.\n"
"\n"
"hbparamter -p <parameter-name>\n"
"       Retrieve the value of cluster parameters.\n"
"       The parameters may be one of the following, however some parameters\n"
"       may not retrieved: hbversion, node, hopfudge, keepalive, deadtime,\n"
"	deadping, warntime, initdead, watchdog, baud, udpport, logfacility,\n"
"	logfile, debugfile, nice_failback, auto_failback, stonith,\n"
"	stonith_host, respawn, rtprio, hbgenmethod, realtime, debug,\n"
"	normalpoll, apiauth, msgfmt.\n\n"
"	get the value of heartbeat parameters.\n"
"	the parameters can be the one of the following, but maybe not all\n"
"	can be actually gotten.\n"
"	hbversion, node, hopfudge, keepalive, deadtime, deadping,\n"
"	warntime, initdead, watchdog, baud, udpport, logfacility,\n"
"	logfile, debugfile, nice_failback, auto_failback, stonith,\n"
"	stonith_host, respawn, rtprio, hbgenmethod, realtime, debug,\n"
"	normalpoll, apiauth, msgfmt\n"
"\n";

static const char * cl_status_name = "cl_status";
static gboolean HB_SIGNON = FALSE;

/*
 * The following is to avoid cl_status sleeping forever. This is due to the 
 * hearbeat's abnormal status or even its crash.  
 */
unsigned int DEFAULT_TIMEOUT = 2;
/* the handler of signal SIGALRM */
static void quit(int signum);

int
main(int argc, char ** argv)
{
	ll_cluster_t *hb = NULL;
	int ret_value = 0; 
	int i = -1;
	gboolean GOOD_CMD = FALSE;

	if ((argc == 1) || ( argc == 2 && strncmp(argv[1],"-h",3) == 0)){
		printf("%s",simple_help_screen);
		return 0;
	}

	cl_log_set_entity(cl_status_name);
	/* 
	 * You can open it for debugging conveniently.
	 * When the program is finished formly, all redundant cl_log clauses 
	 * will be removed
	 */ 
	/* cl_log_enable_stderr(TRUE); */
	cl_log_set_facility(LOG_USER);

	cl_log(LOG_INFO, "start:optind: %d  argv[optind+1]: %s", optind, 
		argv[optind+1]);

	/*
	 * To avoid cl_status sleep forever, trigger a timer and dealing with 
	 * signal SIGALRM. This sleep is due to hearbeat's abnormal status or
	 * its crash during cl_status' execution.  
	 */
	 alarm(DEFAULT_TIMEOUT);
	 signal(SIGALRM, quit);
	
	/*
	 * Don't use getopt_long since its portibility is not good.
	 * Why to using long command, because long commands are natural and 
	 * good to be remembered.
	 */
	while ( cmds[++i].name != NULL ) {
		if ( strncmp(argv[1], cmds[i].name, CMDS_MAX_LENGTH) == 0 ) {
			GOOD_CMD = TRUE;
			hb = ll_cluster_new("heartbeat");
			if ( hb == NULL ) {
				return UNKNOWN_ERROR;
			}

			/* cl_log(LOG_INFO, "Signing in with heartbeat."); */
			if (hb->llc_ops->signon(hb, cl_status_name)!= HA_OK) {
				cl_log(LOG_ERR, "Cannot signon with heartbeat");
				cl_log(LOG_ERR, "REASON: %s", 
					hb->llc_ops->errmsg(hb));
				ret_value = 1;
			}else{
				HB_SIGNON = TRUE;
			}
			ret_value = (cmds[i].func)(hb, argc, argv, 
				cmds[i].optstr);
			break;
		}
	}
	if (GOOD_CMD == FALSE) {
		cl_log(LOG_ERR, "%s is not a correct sub-command.", argv[1]);
		ret_value = PARAMETER_ERROR;
	}

	cl_log(LOG_INFO, "End: optind: %d  argv[optind+1]: %s", optind, 
		argv[optind+1]);

	if (HB_SIGNON == TRUE) {
		if (hb->llc_ops->signoff(hb) != HA_OK) {
			cl_log(LOG_ERR, "Cannot sign off from heartbeat.");
			cl_log(LOG_ERR, "REASON: %s\n", hb->llc_ops->errmsg(hb));
			/* Comment it to avoid to mask the subcommand's return
			 * ret_value = UNKNOWN_ERROR;
			 */
		}
	}
	if (hb != NULL) {
		if (hb->llc_ops->delete(hb) != HA_OK) {
			cl_log(LOG_ERR, "Cannot delete API object.");
			cl_log(LOG_ERR, "REASON: %s", hb->llc_ops->errmsg(hb));
			/* Comment it to avoid to mask the subcommand's return
			 * ret_value = UNKNOWN_ERROR;
			 */
		}
	}

	cl_log(LOG_ERR ,"return value:%d", ret_value);
	return ret_value;
}

static int 
hbstatus(ll_cluster_t *hb, int argc, char ** argv, const char * optstr)
{
	/* Is it ok to judge if heartbeat is running via signon status? */
	if ( HB_SIGNON == TRUE ) {
		printf("Heartbeat is running on this machine.\n");
		return 0;
	} else {
		printf("Heartbeat is stopped on this machine.\n");
		return 1;
	}
}

static int 
listnodes(ll_cluster_t *hb, int argc, char ** argv, const char * optstr)
{
	gboolean LIST_ALL = TRUE,
		 ONLY_LIST_PING = FALSE, 
		 ONLY_LIST_NORAMAL = FALSE;
	int option_char;
	const char * node, * type;

	do {
		option_char = getopt(argc, argv, optstr);

		if (option_char == -1) {
			break;
		}

		switch (option_char) {
			case 'm':
				FOR_HUMAN_READ = TRUE;
				break;

			case 'p':
				ONLY_LIST_PING = TRUE;
				LIST_ALL = FALSE;	
				break;

			case 'n':
				ONLY_LIST_NORAMAL = TRUE;
				LIST_ALL = FALSE;	
				break;

			default:
				cl_log(LOG_ERR, "Error:getopt returned" 
					"character code %c.", option_char);
				return PARAMETER_ERROR;
		}
	} while (1);

	if (hb->llc_ops->init_nodewalk(hb) != HA_OK) {
			cl_log(LOG_ERR, "Cannot start node walk.");
			cl_log(LOG_ERR, "REASON: %s", hb->llc_ops->errmsg(hb));
			return UNKNOWN_ERROR;
	}

	if ( (ONLY_LIST_NORAMAL == TRUE) && (ONLY_LIST_PING == TRUE) ) {
		LIST_ALL = TRUE;
	}
	if (FOR_HUMAN_READ == TRUE) {
		printf("The nodes are as follow:\n");
	}
	while ((node = hb->llc_ops->nextnode(hb))!= NULL) {
		if ( LIST_ALL == TRUE ) {
			printf("%s\n", node);
		} else {
			type = hb->llc_ops->node_type(hb, node);
			if (( ONLY_LIST_NORAMAL == TRUE )
			      && (strncmp(type, "normal", 8)==0)) {
				printf("%s\n", node);
			} else if (( ONLY_LIST_PING == TRUE ) 
				    && (strncmp(type, "ping", 8)==0)) {
				printf("%s\n", node);
			}
		} 
	}

	if (hb->llc_ops->end_nodewalk(hb) != HA_OK) {
		cl_log(LOG_ERR, "Cannot end node walk.");
		cl_log(LOG_ERR, "REASON: %s", hb->llc_ops->errmsg(hb));
		return UNKNOWN_ERROR;
	}
	return 0;
}

/* Map string std_output to return value ? 
 * Active
 */
static int 
nodestatus(ll_cluster_t *hb, int argc, char ** argv, const char * optstr)
{
	const char * status;
	int ret = UNKNOWN_ERROR;

	if ( general_simple_opt_deal(argc, argv, optstr) < 0 ) {
		/* There are option errors */
		return PARAMETER_ERROR;
	};

	if (argc <= optind+1) {
		fprintf(stderr, "No enough parameter.\n");
		return PARAMETER_ERROR;
	}

	cl_log(LOG_INFO, "optind: %d   argv[optindex+1]: %s", optind, 
		argv[optind+1]);
	status = hb->llc_ops->node_status(hb, argv[optind+1]);
	if ( status == NULL ) {
		fprintf(stderr, "Error. May be due to incorrect node name\n");
		return PARAMETER_ERROR;
	}
	if (FOR_HUMAN_READ == TRUE) {
		printf("The cluster nodes %s is %s\n", argv[optind+1], status);
	} else {
		printf("%s\n", status);
	}

       if ( strncmp(status, "active", 6) == 0 
       ||      strncmp(status, "up", 3) == 0) {
		ret = OK;   /* the node is active */
	} else {
		ret = NORMAL_FAIL;  /* the status = "dead" */
	}
	/* Should there be other status? According the comment in heartbeat.c
	   three status: active, up, down. But it's not like that. 
	*/
	
	return ret;
}

/* Map string std_output to return value ? No
 * NORMAL PING UNKNOWN
 */
static int 
nodetype(ll_cluster_t *hb, int argc, char ** argv, const char * optstr)
{
	const char * type;

	if ( general_simple_opt_deal(argc, argv, optstr) < 0 ) {
		/* There are option errors */
		return PARAMETER_ERROR;
	};

	if (argc <= optind+1) {
		fprintf(stderr, "No enough parameter.\n");
		return PARAMETER_ERROR;
	}

	cl_log(LOG_INFO, "optind: %d   argv[optindex+1]: %s", optind, 
		argv[optind+1]);
	type = hb->llc_ops->node_type(hb, argv[optind+1]);
	if ( type == NULL ) {
		fprintf(stderr, "Error. May be due to incorrect node name\n");
		return PARAMETER_ERROR;
	}
	if (FOR_HUMAN_READ == TRUE) {
		printf("The node %s's type: %s\n", argv[optind+1], type);
	} else {
		printf("%s\n", type);
	}
	
	return 0;
}

static int 
listhblinks(ll_cluster_t *hb, int argc, char ** argv, const char * optstr)
{
	const char * intf;

	if ( general_simple_opt_deal(argc, argv, optstr) < 0 ) {
		/* There are option errors */
		return PARAMETER_ERROR;
	};

	if (argc <= optind+1) {
		fprintf(stderr, "No enough parameter.\n");
		return PARAMETER_ERROR;
	}

	cl_log(LOG_INFO, "optind: %d   argv[optindex+1]: %s", optind, 
		argv[optind+1]);

	if (hb->llc_ops->init_ifwalk(hb, argv[optind+1]) != HA_OK) {
		cl_log(LOG_ERR, "Cannot start heartbeat link interface walk.");
		cl_log(LOG_ERR, "REASON: %s", hb->llc_ops->errmsg(hb));
		return UNKNOWN_ERROR;
	}

	if (FOR_HUMAN_READ == TRUE) {
		printf("\tthis node have the following heartbeat links:\n");
	}

	while ((intf = hb->llc_ops->nextif(hb))) {
		printf("\t%s\n", intf); 
	}

	if (hb->llc_ops->end_ifwalk(hb) != HA_OK) {
		cl_log(LOG_ERR, "Cannot end heartbeat link interface walk");
		cl_log(LOG_ERR, "REASON: %s", hb->llc_ops->errmsg(hb));
		return UNKNOWN_ERROR;
	}

	return 0;
}

static int 
hblinkstatus(ll_cluster_t *hb, int argc, char ** argv, const char * optstr)
{
	const char * if_status;	
	int 	ret = UNKNOWN_ERROR;

	if ( general_simple_opt_deal(argc, argv, optstr) < 0 ) {
		/* There are option errors */
		return PARAMETER_ERROR;
	};

	if (argc <= optind+2) {
		fprintf(stderr, "No enough parameter.\n");
		return PARAMETER_ERROR;
	}

	if_status = hb->llc_ops->if_status(hb, argv[optind+1], argv[optind+2]);
	if (if_status == NULL) { /* Shoud be error ? */
		cl_log(LOG_ERR, "Cannot get heartbeat link status");
		cl_log(LOG_ERR, "REASON: %s", hb->llc_ops->errmsg(hb));
		return UNKNOWN_ERROR;
	}

	if (FOR_HUMAN_READ == TRUE) {
		printf("The node %s's heartbeat link %s is %s\n", 
			argv[optind+1], argv[optind+2], if_status);
	} else {
		printf("%s\n", if_status);
	}

	if ( strncmp(if_status, "up", 3) == 0 ) {
		ret = OK; /* the link is up */
	} else {
		ret = NORMAL_FAIL; /* the link is dead */
	}

	return ret;
}

static int 
clientstatus(ll_cluster_t *hb, int argc, char ** argv, const char * optstr)
{
	/* Default value, its unit is milliseconds */
	int timeout = 100;  
	const char * cstatus;
	int ret = UNKNOWN_ERROR;

	if ( general_simple_opt_deal(argc, argv, optstr) < 0 ) {
		/* There are option errors */
		return PARAMETER_ERROR;
	};

	if (argc <= optind+2) {
		fprintf(stderr, "No enough parameter.\n");
		return PARAMETER_ERROR;
	}

	if ( argc > optind+3 ) {
		timeout = atoi(argv[optind+3]);
	}

	cstatus = hb->llc_ops->client_status(hb, argv[optind+1], 
		argv[optind+2], timeout);
	if (cstatus == NULL) { /* Error */
		cl_log(LOG_ERR, "Cannot get client %s's status", argv[optind+2]);
		cl_log(LOG_ERR, "REASON: %s", hb->llc_ops->errmsg(hb));
		return UNKNOWN_ERROR;
	} else {
		/* online, offline, join, leave */
		printf("%s\n", cstatus);
	}

	if ( strncmp(cstatus, "online", 6) == 0 ) {
		ret = OK; 
	} else if ( strncmp(cstatus, "offline", 7) == 0 ) {
		ret = NORMAL_FAIL; /* the client is offline */
	} else if ( strncmp(cstatus, "join", 4) == 0 ) {
		ret = 2;
	} else if ( strncmp(cstatus, "leave", 5) == 0 ) {
		ret = 3;
	}
	
	return ret;
}

static int 
rscstatus(ll_cluster_t *hb, int argc, char ** argv, const char * optstr)
{
	const char * rstatus;

	if ( general_simple_opt_deal(argc, argv, optstr) < 0 ) {
		/* There are option errors */
		return PARAMETER_ERROR;
	};

	rstatus = hb->llc_ops->get_resources(hb);
	if ( rstatus == NULL ) {
		cl_log(LOG_ERR, "Cannot get client %s's status", argv[optind+2]);
		cl_log(LOG_ERR, "REASON: %s", hb->llc_ops->errmsg(hb));
		return UNKNOWN_ERROR;
	}
	if (FOR_HUMAN_READ == TRUE) {
		printf("This node is holding %s resources.\n", rstatus);
	} else {
		printf("%s", rstatus);
	}
	return 0;
}

static int
hbparameter(ll_cluster_t *hb, int argc, char ** argv, const char * optstr)
{
	int option_char;
	int ret_value = 0;
	gchar * paramname = NULL;
	do {
		option_char = getopt(argc, argv, optstr);

		if (option_char == -1) {
			break;
		}

		switch (option_char) {
			case 'm':
				FOR_HUMAN_READ = TRUE;
				break;

			case 'p':
				if (optarg) {
					paramname = g_strdup(optarg);
					cl_log(LOG_ERR,"parameter: %s", 
						paramname);
				}
				break;

			default:
				cl_log(LOG_ERR, "Error:getopt returned" 
					"character code %c.", option_char);
				if (paramname != NULL) {
					g_free(paramname);
				}
				return PARAMETER_ERROR;
		}
	} while (1);

	cl_log(LOG_INFO,"paramname: %s", paramname);
	if ( paramname != NULL ) {
		char * pvalue;
		pvalue = hb->llc_ops->get_parameter(hb, paramname);
		if (pvalue == NULL) {
			cl_log(LOG_ERR, "Cannot get parameter %s's value", 
				argv[optind+2]);
			cl_log(LOG_ERR, "REASON: %s", hb->llc_ops->errmsg(hb));
			ret_value = NORMAL_FAIL;
		} else {
			if (FOR_HUMAN_READ == TRUE) {
				printf("Heartbeat parameter %s's value: %s\n",
					 paramname, pvalue);
			} else {
				printf("%s\n",pvalue);
			}
			free(pvalue);
		}
		g_free(paramname);
	} else {
		cl_log(LOG_ERR, "Need parameter's name");
		ret_value = PARAMETER_ERROR;
	}

	return ret_value;
}

static int 
general_simple_opt_deal(int argc, char ** argv, const char * optstr)
{
	int option_char;
	do {
		option_char = getopt(argc, argv, optstr);

		if (option_char == -1) {
			break;
		}

		switch (option_char) {
			case 'm':
				FOR_HUMAN_READ = TRUE;
				break;

			default:
				cl_log(LOG_ERR, "Error:getopt returned"
					"character code %c.", option_char);
				return PARAMETER_ERROR;
		}
	} while (1);
	return 0;
}

static int
test(ll_cluster_t *hb, int argc, char ** argv, const char * optstr)
{
	printf("Dead time: %d\n keeplive time: %d\n mynodeid: %s\n rsc: %s\n",
		(int)hb->llc_ops->get_deadtime(hb), 
		(int)hb->llc_ops->get_keepalive(hb),
		hb->llc_ops->get_mynodeid(hb), hb->llc_ops->get_resources(hb));
	return 0;
}

/* the handler of signal SIGALRM */
static void
quit(int signum)
{
	if ( signum == SIGALRM ) {
		exit(TIMEOUT);
	}
}
